### ISSUE TITLE
Pyarmor in Docker: ERROR    invalid license token, try to run `pyarmor reg` to register license again

### ISSUE THREAD

**Issue by goraje** [dashingsoft/pyarmor#1542]
Hi, I've been trying to run Pyarmor from a Docker container, but even though the registration process was fixed with Pyarmor 8.4.0 when I run code obfuscation with `pyarmor gen` I'm getting the following error:
```
root@4b279a129f6b:/# pyarmor reg pyarmor-device-regfile-6005.7.zip
INFO     Python 3.11.6
INFO     Pyarmor 8.4.0 (trial), 000000, non-profits
INFO     Platform linux.aarch64
INFO     register "pyarmor-device-regfile-6005.7.zip"
INFO     machine id in group license: m98a1e7f27fc612f428043fb2edb27122
INFO     got machine id: m47f31be7ced2f9161c5221e56364fac2
INFO     got machine id: l47f31be7ced2f9161c5221e56364fac2
INFO     got machine id: i98444de277e1b3dd9b17d2181631a73c
INFO     got machine id: k1a298b2ce465500b355f6c9d777b845f
INFO     got machine id: g1a298b2ce465500b355f6c9d777b845f
INFO     got machine id: bb6889a725ecfcc7acd18b1dd249f4df9
INFO     no machine id matchs this group license
INFO     take this machine as docker container, and connect to docker host for authentication...
INFO     got docker host machine id: m98a1e7f27fc612f428043fb2edb27122
INFO     got docker host machine id: l98a1e7f27fc612f428043fb2edb27122
INFO     got docker host machine id: i98a1e7f27fc612f428043fb2edb27122
INFO     got docker host machine id: kc6eb2c01acd9320d96ee5992197d6a44
INFO     got docker host machine id: gc6eb2c01acd9320d96ee5992197d6a44
INFO     got docker host machine id: b28c5a05ba0fa8e759f3e172660d31c04
INFO     This license registration information:

License Type    : pyarmor-group
License No.     : pyarmor-vax-006005
License To      : *****
License Product : ******

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation

root@4b279a129f6b:/# pyarmor gen foo.py
INFO     Python 3.11.6
INFO     Pyarmor 8.4.0 (group), 006005, ******
INFO     Platform linux.aarch64
INFO     search inputs ...
INFO     find script foo.py
INFO     find 1 top resources
ERROR    invalid license token, try to run `pyarmor reg` to register license again
```

and `pyarmor.error.log` still showing:

```
266 MainProcess 2023-10-18 06:58:09,818
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/__main__.py", line 712, in main
    main_entry(sys.argv[1:])
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/__main__.py", line 700, in main_entry
    return args.func(ctx, args)
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/__main__.py", line 230, in cmd_gen
    builder.process(options)
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/generate.py", line 147, in process
    Pytransform3.pre_build(self.ctx)
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/core/__init__.py", line 117, in pre_build
    m = Pytransform3.init(ctx)
        ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/core/__init__.py", line 97, in init
    m.init_ctx(ctx)
RuntimeError: invalid license token, try to run `pyarmor reg` to register license again
```
If I repeat the `pyarmor reg` registration procedure exactly the same thing happens. I'm working on an Apple silicon MacBook Pro. 
One more thing that may or may not be of importance in this issue is that I can see the communication between the `pyarmor-auth` service and the container when I run `pyarmor reg` (I can see packets received and packets sent in the log of `pyarmor-auth`), whereas there is no indication of any communication when I run `pyarmor gen`.

**Comment by jondy** [comment]
What's the output of `pyarmor -v` in docker container?

**Comment by jondy** [comment]
Try to start docker container with extra options `--add-host=host.docker.internal:host-gateway`

**Comment by jondy** [comment]
And is there `/dev/disk` in docker container?

**Comment by goraje** [comment]
Hi, I did what you instructed and this is the output. The problem still persists.
```
root@446598bcab51:/# pyarmor -v
Pyarmor 8.4.0 (group), 006005, RADKit

License Type    : pyarmor-group
License No.     : pyarmor-vax-006005
License To      : *****
License Product : ******

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation

INFO     got docker host machine id: m98a1e7f27fc612f428043fb2edb27122
INFO     got docker host machine id: l98a1e7f27fc612f428043fb2edb27122
INFO     got docker host machine id: i98a1e7f27fc612f428043fb2edb27122
INFO     got docker host machine id: kc6eb2c01acd9320d96ee5992197d6a44
INFO     got docker host machine id: gc6eb2c01acd9320d96ee5992197d6a44
INFO     got docker host machine id: b28c5a05ba0fa8e759f3e172660d31c04
root@446598bcab51:/# echo "print('TEST')" > foo.py
root@446598bcab51:/# pyarmor gen foo.py
INFO     Python 3.11.6
INFO     Pyarmor 8.4.0 (group), 006005, RADKit
INFO     Platform linux.aarch64
INFO     search inputs ...
INFO     find script foo.py
INFO     find 1 top resources
ERROR    invalid license token, try to run `pyarmor reg` to register license again
```

**Comment by goraje** [comment]
No `/dev/disk` is not listed as one of the devices
```
root@7b51114ec541:/# ls /dev/
console  fd  full  mqueue  null  ptmx  pts  random  shm  stderr  stdin	stdout	tty  urandom  zero
```

**Comment by jondy** [comment]
How about `$ tracert host.docker.internal`?

**Comment by jondy** [comment]
It seems it should be `traceroute` not `tracert`. If not installed, try to install it by `apt install traceroute` in docker container

**Comment by jondy** [comment]
And what's the output of `pyarmor-auth`? It listens on IPv4 or IPv6 address, if it's IPv6, this will be problem.

**Comment by goraje** [comment]
Here' the `traceroute` output
```
root@0b5903d41cc2:/# traceroute host.docker.internal
traceroute to host.docker.internal (192.168.5.2), 30 hops max, 60 byte packets
 1  172.17.0.1 (172.17.0.1)  0.067 ms  0.012 ms  0.008 ms
 2  host.docker.internal (192.168.5.2)  0.727 ms  0.691 ms  0.666 ms
 3  * * *
 4  * * *
 5  * * *
 6  * * *
 7  * * *
 8  * * *
 9  * * *
10  * * *
11  * * *
12  * * *
13  * * *
14  * * *
15  * * *
16  * * *
17  * * *
18  * * *
19  * * *
20  * * *
21  * * *
22  * * *
23  * * *
24  * * *
25  * * *
26  * * *
27  * * *
28  * * *
29  * * *
30  * * *
```
and also maybe this could be useful:
```
root@0b5903d41cc2:/# telnet host.docker.internal 29092
Trying 192.168.5.2...
Connected to host.docker.internal.
Escape character is '^]'.
```

**Comment by goraje** [comment]
But the registration proceeds without fail so it seems that the connectivity to the auth service is there. There's just no call being made to it when I run `pyarmor gen` as far as I can tell.

**Comment by jondy** [comment]
Try this solution, make sure `host.docker.internal` parsed to `172.17.0.1` in both of docker container and host

For example, adding one line in `/etc/hosts`
```
host.docker.internal 172.17.0.1
```

**Comment by goraje** [comment]
Hi, actually in my case on a Mac in the Docker container the `host.docker.internal` entry resolves to 192.168.5.2 and when I set it to the same value in `/etc/hosts` of the macOS host it doesn't change or fix the behavior with `pyarmor gen` (`pyarmor reg` still works). Setting the value to 172.17.0.1 is a no go because this network does not exist on the host.

**Comment by jondy** [comment]
Well, how about start docker with `--add-host=host.docker.internal:172.17.0.1`

**Comment by jondy** [comment]
If `host.docker.internal` isn't in same network of docker container, it will not sent auth request to docker host. This is the problem.

**Comment by goraje** [comment]
Then the registration fails
```
root@025a348e78d0:/# pyarmor reg pyarmor-device-regfile-6005.7.zip
INFO     Python 3.11.6
INFO     Pyarmor 8.4.0 (trial), 000000, non-profits
INFO     Platform linux.aarch64
INFO     register "pyarmor-device-regfile-6005.7.zip"
INFO     machine id in group license: m98a1e7f27fc612f428043fb2edb27122
INFO     got machine id: m64bf28cba3e2661680ccc54b3cd92b41
INFO     got machine id: l64bf28cba3e2661680ccc54b3cd92b41
INFO     got machine id: ia3e940b241521a711b8961de85aaedfb
INFO     got machine id: k6ff0851f0ca3db1f69c2eae5d43688e2
INFO     got machine id: g6ff0851f0ca3db1f69c2eae5d43688e2
INFO     got machine id: b7e263a34ce45701ac80be26b42806151
INFO     no machine id matchs this group license
INFO     take this machine as docker container, and connect to docker host for authentication...
INFO     could not get docker host machine id

if this machine is docker container, please run command `pyarmor-auth` in docker host, and try it again

otherwise please generate new group device license for this machine

more information please check section "using group license" in documentation "how-to register" guide

ERROR    this group device license is not for this machine
```
and there is no connectivity between the host and the container

**Comment by goraje** [comment]
> If `host.docker.internal` isn't in same network of docker container, it will not sent auth request to docker host. This is the problem.

On macOS and Windows for the matter it won't be the case, because contrary to Linux you cannot run Docker natively. Instead you run it on the side in a Linux VM that resides within some kind of a private network (usually 192.*.*.*). The Docker network 172.*.*.* will exist only on the Linux VM but will not be present on your macOS/Windows host. Instead the `docker.sock` from the VM is passed to the host OS to allow for interaction via Docker's CLI and communication between the containers and the host is port forwarding-based.

**Comment by jondy** [comment]
So is it possible to run `pyarmor-auth` on the Linux VM which is running in docker host (windows/MacOs)?

**Comment by jondy** [comment]
What's the output of `ifconfig -a` in docker host (MacOS)?

**Comment by goraje** [comment]
And just to reiterate what I've said before. The container can send requests to `host.docker.interal:29092` - the `pyarmor-auth` service and receive responses (`pyarmor reg` both sent and received a response).

**Comment by jondy** [comment]
> and this is the man-in-between (the Linux VM I mentioned before)

Is it possible to run `pyarmor-auth` in this Linux VM? I have no more idea about docker in MacOs/Windows.

**Comment by jondy** [comment]
> And just to reiterate what I've said before. The container can send
>  requests to  host.docker.interal:29092  - the  pyarmor-auth  service
>  and receive responses ( pyarmor reg  both sent and received a      
>  response).                                                         

So what's the ip address of `host.docker.internal` in docker container?It's `172.x...` or `192.x...`

**Comment by goraje** [comment]
I've tried it out also setting host.docker.internal to 172.17.0.1 on both the Linux VM and the container. The problem still persists and this is the log from `pyarmor-auth`.
```
colima-pyarmor:/Users/jgora$ pyarmor-auth --debug pyarmor-device-regfile-6005.8.zip
2023-10-19 14:50:45,961: work path: /home/jgora.linux/.pyarmor/docker
2023-10-19 14:50:45,963: register "pyarmor-device-regfile-6005.8.zip"
2023-10-19 14:50:45,967: extracting license.lic
2023-10-19 14:50:45,967: extracting .pyarmor_capsule.zip
2023-10-19 14:50:45,968: machine id in group license: mec58608fff941488f5a72f812f3203c2
2023-10-19 14:50:45,968: got machine id: mec58608fff941488f5a72f812f3203c2
2023-10-19 14:50:45,968: this machine id matchs group license
2023-10-19 14:50:45,968: extracting tokens/mec58608fff941488f5a72f812f3203c2
2023-10-19 14:50:45,970: machine id: [b'mec58608fff941488f5a72f812f3203c2', b'lec58608fff941488f5a72f812f3203c2', b'i02dcbedfc6e0dd4f6b2d59a758fa9f15', b'kba0ba013eeb9a873be30a01d64da31e7', b'gba0ba013eeb9a873be30a01d64da31e7', b'b310417f33732ca2f9ce1e01e0b05c416']
2023-10-19 14:50:45,970: listen container auth request on 0.0.0.0:29092
2023-10-19 14:51:02,989: receive request from ('172.17.0.2', 53510)
2023-10-19 14:51:02,989: request data (64): b'PADHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
2023-10-19 14:51:02,989: send auth result to ('172.17.0.2', 53510)
2023-10-19 14:51:02,989: response data (204): b'mec58608fff941488f5a72f812f3203c2\nlec58608fff941488f5a72f812f3203c2\ni02dcbedfc6e0dd4f6b2d59a758fa9f15\nkba0ba013eeb9a873be30a01d64da31e7\ngba0ba013eeb9a873be30a01d64da31e7\nb310417f33732ca2f9ce1e01e0b05c416\x00'
2023-10-19 14:51:09,439: receive request from ('172.17.0.2', 56422)
2023-10-19 14:51:09,439: request data (64): b'PADHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
2023-10-19 14:51:09,439: send auth result to ('172.17.0.2', 56422)
2023-10-19 14:51:09,439: response data (204): b'mec58608fff941488f5a72f812f3203c2\nlec58608fff941488f5a72f812f3203c2\ni02dcbedfc6e0dd4f6b2d59a758fa9f15\nkba0ba013eeb9a873be30a01d64da31e7\ngba0ba013eeb9a873be30a01d64da31e7\nb310417f33732ca2f9ce1e01e0b05c416\x00'
2023-10-19 14:51:16,028: receive request from ('172.17.0.2', 56432)
2023-10-19 14:51:16,028: request data (32): b'PADK\x00\x00\x00\x00\x00\x00\x00\x00mlhdjeaemmbldfdfibgk'
Segmentation fault
```
Requests/responses:
1. after `pyarmor reg` command
2. after `pyarmor -v` command in the container
3. after `pyarmor gen` in container - causes the segmentation fault

**Comment by goraje** [comment]
> But I think `host.docker.internal` could be configured to `172.17.0.1` in docker container by editing `/etc/hosts` in docker container.
> 
> Even it may not be changed in docker host.

When I worked directly on the Linux VM I've set it to 172.17.0.1 on both the container and the Linux VM

**Comment by jondy** [comment]
Or just for a test, patch `/path/to/pyarmor/cli/register.py` in docker container, replace `host.docker.internal` with `172.17.0.1`, then run `pyarmor reg xxx`.

If docker host `pyarmor-auth` could receive request and send response, this will be one solution.

**Comment by goraje** [comment]
> Could you run `pyarmor gen key --bind-data test -e 3` in docker host? Just test `generate_runtime_key` crash or not.

Of course. It seems that it works just fine on the Docker host (Linux VM) itself.
```
colima-pyarmor:/Users/jgora$ pyarmor gen key --bind-data test -e 3
INFO     Python 3.10.13
INFO     Pyarmor 8.4.0 (group), 006005, ******
INFO     Platform linux.aarch64
INFO     start to generate outer runtime key "pyarmor.rkey"
INFO     write dist/pyarmor.rkey
INFO     generate outer runtime key OK
```

**Comment by goraje** [comment]
By editing 'pyarmor/cli/docker.py` and `pyarmor/cli/generate.py` I've managed to establish that the `Segmentation fault` occurs in `Builder`'s `generate_runtime_key` when it invokes `Pytransform3.generate_runtime_key(self.ctx, outer)`. I don't know if this is helpful but when I execute `pyarmor gen key --bind-data test -e 3` inside of the Docker container the `ctx` object received by the builder looks as follows:
```
{'alias_suffix': '{0}-{1}',
 'bootstrap_template': '# Pyarmor $rev, $timestamp\n'
                       'from $package import __pyarmor__\n'
                       '__pyarmor__(__name__, $path, $code)\n',
 'cfg': <configparser.ConfigParser object at 0xffff9df182e0>,
 'cmd_options': {'user_data': 'goejobmjndofhlbkknme'},
 'encoding': None,
 'extra_libs': {},
 'extra_resources': [],
 'global_path': '/home/jgora.linux/.pyarmor/docker/config',
 'home_path': '/home/jgora.linux/.pyarmor/docker',
 'input_paths': [],
 'local_path': '.pyarmor',
 'module_builtins': set(),
 'module_relations': {},
 'module_types': {},
 'obfuscated_modules': set(),
 'outputs': [],
 'plugins': [],
 'reg_path': '/home/jgora.linux/.pyarmor/docker',
 'resources': [],
 'rft_auto_excludes': {'super'},
 'rft_export_names': set(),
 'rft_transform_op': '?',
 'runtime_key': None,
 'runtime_keyfile': '.pyarmor.ikey',
 'variable_types': {}}
```

**Comment by jondy** [comment]
This patch may fix segmentfault issue
```patch
--- a/src/cli/docker.py
+++ b/src/cli/docker.py
@@ -53,7 +53,8 @@ class DockerAuthHandler(socketserver.BaseRequestHandler):
     def generate_runtime_key(self, userdata):
         ctx = CONFIG['ctx']
         ctx.cmd_options['user_data'] = userdata
-        return Builder(ctx).generate_runtime_key()
+        Pytransform3._pytransform3.init_ctx(ctx)
+        return Pytransform3.generate_runtime_key(ctx)
 ```

**Comment by jondy** [comment]
A quick fixed version 8.4.1 has been released for this issue

**Comment by jondy** [comment]
Great!

It's better to use flag `22`, for others (11 to 21), they may not work in some cases. For example, there are multiple physical network cards.
